pkg/utils/security.go

```diff
@@ -17,41 +17,170 @@ import (
 	"crypto/tls"
 	"crypto/x509"
 	"io/ioutil"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"strings"
 
 	"github.com/pingcap/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
 )
 
+// TLS ...
+type TLS struct {
+	inner  *tls.Config
+	client *http.Client
+	url    string
+}
+
 // ToTLSConfig generates tls's config.
-func ToTLSConfig(SSLCA, SSLCert, SSLKey string) (*tls.Config, error) {
-	var tlsConfig *tls.Config
-	if len(SSLCA) != 0 {
-		var certificates = make([]tls.Certificate, 0)
-		if len(SSLCert) != 0 && len(SSLKey) != 0 {
-			// Load the client certificates from disk
-			certificate, err := tls.LoadX509KeyPair(SSLCert, SSLKey)
-			if err != nil {
-				return nil, errors.Errorf("could not load client key pair: %s", err)
-			}
-			certificates = append(certificates, certificate)
-		}
+func ToTLSConfig(caPath, certPath, keyPath string) (*tls.Config, error) {
+	return ToTLSConfigWithVerify(caPath, certPath, keyPath, nil)
+}
 
-		// Create a certificate pool from the certificate authority
-		certPool := x509.NewCertPool()
-		ca, err := ioutil.ReadFile(SSLCA)
+// ToTLSConfigWithVerify constructs a `*tls.Config` from the CA, certification and key
+// paths, and add verify for CN.
+//
+// If the CA path is empty, returns nil.
+func ToTLSConfigWithVerify(caPath, certPath, keyPath string, verifyCN []string) (*tls.Config, error) {
+	if len(caPath) == 0 {
+		return nil, nil
+	}
+
+	// Load the client certificates from disk
+	var certificates []tls.Certificate
+	if len(certPath) != 0 && len(keyPath) != 0 {
+		cert, err := tls.LoadX509KeyPair(certPath, keyPath)
 		if err != nil {
-			return nil, errors.Errorf("could not read ca certificate: %s", err)
+			return nil, errors.Annotate(err, "could not load client key pair")
 		}
+		certificates = []tls.Certificate{cert}
+	}
 
-		// Append the certificates from the CA
-		if !certPool.AppendCertsFromPEM(ca) {
-			return nil, errors.New("failed to append ca certs")
-		}
+	// Create a certificate pool from CA
+	certPool := x509.NewCertPool()
+	ca, err := ioutil.ReadFile(caPath)
+	if err != nil {
+		return nil, errors.Annotate(err, "could not read ca certificate")
+	}
 
-		tlsConfig = &tls.Config{
-			Certificates: certificates,
-			RootCAs:      certPool,
+	// Append the certificates from the CA
+	if !certPool.AppendCertsFromPEM(ca) {
+		return nil, errors.New("failed to append ca certs")
+	}
+
+	tlsCfg := &tls.Config{
+		Certificates: certificates,
+		RootCAs:      certPool,
+		ClientCAs:    certPool,
+		NextProtos:   []string{"h2", "http/1.1"}, // specify `h2` to let Go use HTTP/2.
+	}
+
+	checkCN := make(map[string]struct{})
+	for _, cn := range verifyCN {
+		cn = strings.TrimSpace(cn)
+		checkCN[cn] = struct{}{}
+	}
+	if len(checkCN) != 0 {
```

Comment:
move lines 80-84 inside the `if` branch.

```suggestion
	if len(verifyCN) != 0 {
		checkCN := make(map[string]struct{})
		for _, cn := range verifyCN {
			cn = strings.TrimSpace(cn)
			checkCN[cn] = struct{}{}
		}
```

Reply:
refine in dea0b3a